OrbitSsh/DirectTcpIpChannelManager.cpp

```diff
@@ -31,107 +30,77 @@ DirectTcpIpChannelManager::DirectTcpIpChannelManager(
 // * kWaitRemoteClosed: Waiting for the remote server to close the channel. Once
 // it is closed the state jumps back to kNotInitialized and tries to establish a
 // new connection.
-DirectTcpIpChannelManager::State DirectTcpIpChannelManager::Tick() {
-  switch (state_) {
-    case State::kNotInitialized: {
-      if (Channel::CreateDirectTcpIp(session_ptr_, third_party_host_,
-                                     third_party_port_,
-                                     &channel_) == ResultType::kSuccess) {
-        state_ = State::kRunning;
-        LOG("Direct Tcp Channel Connected");
-      }
-      break;
-    }
-    case State::kRunning: {
-      break;
-    }
-    case State::kSentEOFToRemote: {
-      if (channel_->WaitRemoteEOF() != ResultType::kAgain) {
-        state_ = State::kRemoteSentEOFBack;
-      }
-      break;
-    }
-    case State::kRemoteSentEOFBack: {
-      if (channel_->Close() != ResultType::kAgain) {
-        state_ = State::kWaitRemoteClosed;
-      }
-      break;
-    }
-    case State::kWaitRemoteClosed: {
-      if (channel_->WaitClosed() != ResultType::kAgain) {
-        channel_ = std::nullopt;
-        state_ = State::kNotInitialized;
-      }
-      break;
-    }
+outcome::result<DirectTcpIpChannelManager::State>
+DirectTcpIpChannelManager::Tick() {
+  if (state_ == State::kNotInitialized) {
+    OUTCOME_TRY(Initialize());
+  }
+  if (state_ != State::kRunning) {
+    OUTCOME_TRY(Close());
   }
-
   return state_;
 }
-
-ResultType DirectTcpIpChannelManager::WriteBlocking(const std::string& text) {
-  if (state_ == State::kRunning) return channel_->WriteBlocking(text);
-
-  ERROR("Unable to write to DirectTcpIpChannelManager, state not running");
-  return ResultType::kError;
-}
-
-ResultType DirectTcpIpChannelManager::Read(std::string* result) {
-  if (state_ != State::kRunning) {
-    ERROR("Unable to write to DirectTcpIpChannelManager, state not running");
-    return ResultType::kError;
+outcome::result<void> DirectTcpIpChannelManager::Initialize() {
+  if (state_ == State::kNotInitialized) {
+    OUTCOME_TRY(channel,
+                Channel::OpenTcpIpTunnel(session_ptr_, third_party_host_,
+                                         third_party_port_));
+    channel_ = std::move(channel);
+    state_ = State::kRunning;
+    LOG("Direct Tcp Channel Connected");
   }
+  return outcome::success();
+}
 
-  ResultType read_result = channel_->Read(result);
-
-  if (channel_->GetRemoteEOF()) {
-    state_ = State::kRemoteSentEOFBack;
-    LOG("Direct Tcp Channel Disconnected");
-    return ResultType::kError;
-  }
+outcome::result<void> DirectTcpIpChannelManager::WriteBlocking(
+    std::string_view data) {
+  CHECK(state_ == State::kRunning);
+  return channel_->WriteBlocking(data);
+}
 
-  return read_result;
+outcome::result<std::string> DirectTcpIpChannelManager::Read() {
+  CHECK(state_ == State::kRunning);
+  return channel_->Read();
```

Comment:
In this Read function the case that detects when the channel disconnects is removed. This needs to happen at one point, maybe not here but then in a higher level (tunnel manager)

Reply:
I see. I added a check in TunnelManager.